README
No much to add different from last README, basically the assignment were pretty similar so I just had to tweak for new variables to be read from the test file. Fot the types I created a bitfield so it eas easier to identify the type of object. I will add a drief description of my last assignment READMe just in case the T.A that is gradding the assignment didn't grade my last. This way it would have  abetter understanding of what I did. 
For this assigment I used the base knowledge we gained on C++ from lectures, plus some websites that gave me the template to start using ifstream. For parsing I just used a combination of getline() -> with delimeters, and the substr() function to parse some fields (like the diece variables). This assigment was very helpfull to get started with C++. It gave me experince on C++, integration of C with C++, how prepare the C files for use with C++. I want to add that for time constrains I didn't create a dice object to save the dice variables (base,dice,sides), BUT I did parse them and store them in Ints, also I sotre the color and abil of the npc in int variables that is values is determine by a bit field (this is what was required for the assigment). 




CHANGELOG
///ASSIGMENT 1///
#01/25/2015 (1.01 Dungeon Generation)
-Started the dungeon 
-Created the dungeon representation in a 2-D array [160][96]
-represented the rooms with a struct
-Created the function to initialize the dimensions of a room


#01/26/2015
-implemented create_rooms(), all rooms are place correctly on the map with all rules met
-Corridors implement in vertical and horizontal direction, Used the centro point of two 
consecutive place rooms to create them.
*corridors are experience a incomplete block at the end of output
*Having problems to compile code when tried to create header file

#01/27/2015
-Fixed compilation problem, I was missing a ";" on the header file
-Fixed corridors
///ASIGMENT 1 DONE ///

//ASSIGMENT 2 ///
#01/30/2015
-Started assignment
-Used as template code of FILE I/O that we wrote in class.
-Declare Dungeon array that would represent the dungeon in the binary file
-Started to write to dungeon binary file. Successfully wrote the header "RLG229"

#02/01/2015
-Successfully recorded the map cells in the binary file
-Working in getting the file size and the room coordinates (To be completed)
-It seems all the fields are saved correctly on the file

#02/02/2015
-got working load and save
-work pretty nice
-Clean code, gonna review with professor details and submit
//ASSIGMENT 2 DONE ///

//ASIGMENT 3 //

02/6/2015
- started planning heap min data struct
- Researche wikipedia article to develo dijktra algorithm for finding shortest path

02/10/2015
- Started implementation of heap min, still trying to figure out how sort elements by key
- Created a typedef struct "node_path" that wil hold cost, position, index on heap and previous
node, this will provide data to dijktra algorithm .

02/12/2015
- Almost finished heap_min, still figuring out how implement correctly the decrease_key function
- logic for dijktra almost complete

02/13/2015
- Finally got done with the min heap
- Having a lot of seg faults when trying to test algorithm, will start testing by parts

02/16/2015
- Seg fault are gone, but the map only show one 'P' ('P' is my debug terrain to show the shortest path)
- Logi seem to be fine, sadly it must be a small bug what I have

02/17/2015
-Finally got it working 99% of the time, very few time it outputs a random 'P' not following the path
- I have spent +50 hours in this assignment, so I hope this small bug doesn't affect severe my grade. 

//ASSIGMNET 3 DONE


//ASSIGMENT 4 START///

02/19/2015
- started planning assigment
- Implemented Character_t struct that represent a character on the map

02/20/2015
- Created a modified dijktra algorithm called dijktra_game, to keep turn of character
- Having some problems moving the characters 

02/21/2015
- Telepathic & Smart Monster moving correctly, also PC moving left-and right only 
- Starting implementing non-smart and only telepathic monstes

02/23/2015
- All monsters type implemented correctly, 
- Starting implementing ending of game

02/24/2015
- All implemented correctly, Assignment done.

//Assignment 4 done 

//Assignment 5 
03/01/2015
- starting planning strategy for assignment 
- include initialization ncurses function from lectures

02/01/2015 
- finished character movement 
- finished displaying only 80*24 display 

03/01/2015
- finished Look-mode
- not had time to do stairs part
//Assignment 5 done

//Assignment 6///

03/19/2015
-started assignment
--thinking approach to write info to file
- noticed that all the info from the PC can be saved
using the dungeon structure
-almost finished the writing portion of the assigment

03/20/2015
- had to configure the dugeon writing function, due to the 
stairs
- finished PC data saving, starting NPC storage.
- all the info about the NPC can be getted from the dungeon
structure
- finished NPC portion, starting modifiying reading portion
- reading portion done, starting auto-loading and deletion of 
game file when finished.
- game done. 

//Assigment 6 done

//Assigment 7///

03/30/2015
- created a parser.cpp and parser.h files for the assigment
- started creating a strign isntance to parse the text file 
using the getline() method
- use getenv() function to get to the monster file on the /.rlg229 folder
- Having some problems parsing number from strings

- finalyly got to parse the diece variables from speed,ham, hp, assigment done

03/31/2015

- added variables ints to store the color and abil of the monsters

//Assigment 7 done (In one day :) )

//Start Assignment 8////

04/07/2014 

-start assignmet base on last week assignmet
- replaced/added variable names to accomodate the parser for new text file
- create bit field for the new TYPES
- tested the assignmen (IT WORKS :) 

/////Done assignment 8 /////////////




io.c
#include <unistd.h>
#include <ncurses.h>

#include "io.h"
#include "move.h"

/* We're going to be working in a standard 80x24 terminal, ⁄and, except when *
 * the PC is near the edges, we're going to restrict it to the centermost   *
 * 60x18, defining the center to be at (40,12).  So the PC can be in the    *
 * rectangle defined by the ordered pairs (11, 4) and (70, 21).  When the   *
 * PC leaves this zone, if already on the corresponding edge of the map,    *
 * nothing happens; otherwise, the map shifts by +-40 in x or +- 12 in y,   *
 * such that the PC is in the center 60x18 region.  Look mode will also     *
 * shift by 40x12 blocks.  Thus, the set of all possible dungeon positions  *
 * to correspond with the upper left corner of the dungeon are:             *
 *                                                                          *
 *   ( 0,  0), (40,  0), (80,  0)                                           *
 *   ( 0, 12), (40, 12), (80, 12)                                           *
 *   ( 0, 24), (40, 24), (80, 24)                                           *
 *   ( 0, 36), (40, 36), (80, 36)                                           *
 *   ( 0, 48), (40, 48), (80, 48)                                           *
 *   ( 0, 60), (40, 60), (80, 60)                                           *
 *   ( 0, 72), (40, 72), (80, 72)                                           */

void io_init_terminal(void)
{
  initscr();
  raw();
  noecho();
  curs_set(0);
  keypad(stdscr, TRUE);
}

void io_reset_terminal(void)
{
  endwin();
}

void io_calculate_offset(dungeon_t *d)
{
  d->io_offset[dim_x] = ((d->pc.position[dim_x] - 20) / 40) * 40;
  if (d->io_offset[dim_x] < 0) {
    d->io_offset[dim_x] = 0;
  }
  if (d->io_offset[dim_x] > 80) {
    d->io_offset[dim_x] = 80;
  }
  d->io_offset[dim_y] = ((d->pc.position[dim_y] - 6) / 12) * 12;
  if (d->io_offset[dim_y] < 0) {
    d->io_offset[dim_y] = 0;
  }
  if (d->io_offset[dim_y] > 72) {
    d->io_offset[dim_y] = 72;
  }

#if 0
  uint32_t y;
  uint32_t min_diff, diff;

  min_diff = diff = abs(d->pc.position[dim_x] - 40);
  d->io_offset[dim_x] = 0;
  if ((diff = abs(d->pc.position[dim_x] - 80)) < min_diff) {
    min_diff = diff;
    d->io_offset[dim_x] = 40;
  }
  if ((diff = abs(d->pc.position[dim_x] - 120)) < min_diff) {
    min_diff = diff;
    d->io_offset[dim_x] = 80;
  }

  /* A lot more y values to deal with, so use a loop */

  for (min_diff = 96, d->io_offset[dim_y] = 0, y = 12; y <= 72; y += 12) {
    if ((diff = abs(d->pc.position[dim_y] - (y + 12))) < min_diff) {
      min_diff = diff;
      d->io_offset[dim_y] = y;
    }
  }
#endif
}

void io_update_offset(dungeon_t *d)
{
  int32_t x, y;

  x = (40 + d->io_offset[dim_x]) - d->pc.position[dim_x];
  y = (12 + d->io_offset[dim_y]) - d->pc.position[dim_y];

  if (x >= 30 && d->io_offset[dim_x]) {
    d->io_offset[dim_x] -= 40;
  }
  if (x <= -30 && d->io_offset[dim_x] != 80) {
    d->io_offset[dim_x] += 40;
  }
  if (y >= 8 && d->io_offset[dim_y]) {
    d->io_offset[dim_y] -= 12;
  }
  if (y <= -8 && d->io_offset[dim_y] != 72) {
    d->io_offset[dim_y] += 12;
  }
}

void io_display_huge(dungeon_t *d)
{
  uint32_t y, x;

  clear();
  for (y = 0; y < DUNGEON_Y; y++) {
    for (x = 0; x < DUNGEON_X; x++) {
      if (d->character[y][x]) {
        mvaddch(y, x, d->character[y][x]->symbol);
      } else {
        switch (mapxy(x, y)) {
        case ter_wall:
        case ter_wall_no_room:
        case ter_wall_no_floor:
        case ter_wall_immutable:
          mvaddch(y, x, '#');
          break;
        case ter_floor:
        case ter_floor_room:
        case ter_floor_hall:
        case ter_floor_tentative:
          mvaddch(y, x, '.');
          break;
        case ter_debug:
          mvaddch(y, x, '*');
          break;
        case ter_stairs_up:
          mvaddch(y, x, '<');
          break;
        case ter_stairs_down:
          mvaddch(y, x, '>');
          break;
        default:
 /* Use zero as an error symbol, since it stands out somewhat, and it's *
  * not otherwise used.                                                 */
          mvaddch(y, x, '0');
        }
      }
    }
  }
  refresh();
}

void io_display(dungeon_t *d)
{
  uint32_t y, x;

  clear();
  for (y = 0; y < 24; y++) {
    for (x = 0; x < 80; x++) {
      if (d->character[d->io_offset[dim_y] + y]
                      [d->io_offset[dim_x] + x]) {
        mvaddch(y, x, d->character[d->io_offset[dim_y] + y]
                                  [d->io_offset[dim_x] + x]->symbol);
      } else {
        switch (mapxy(d->io_offset[dim_x] + x,
                      d->io_offset[dim_y] + y)) {
        case ter_wall:
        case ter_wall_no_room:
        case ter_wall_no_floor:
        case ter_wall_immutable:
          mvaddch(y, x, '#');
          break;
        case ter_floor:
        case ter_floor_room:
        case ter_floor_hall:
        case ter_floor_tentative:
          mvaddch(y, x, '.');
          break;
        case ter_debug:
          mvaddch(y, x, '*');
          break;
        case ter_stairs_up:
          mvaddch(y, x, '<');
          break;
        case ter_stairs_down:
          mvaddch(y, x, '>');
          break;
        default:
 /* Use zero as an error symbol, since it stands out somewhat, and it's *
  * not otherwise used.                                                 */
          mvaddch(y, x, '0');
        }
      }
    }
  }
  mvprintw(0, 0, "PC position is (%3d,%2d); offset is (%3d,%2d).",
           d->pc.position[dim_x], d->pc.position[dim_y],
           d->io_offset[dim_x], d->io_offset[dim_y]);


  refresh();
}

void io_look_mode(dungeon_t *d)
{
  int32_t key;

  do {
    if ((key = getch()) == 27 /* ESC */) {
      io_calculate_offset(d);
      io_display(d);
      return;
    }
    
    switch (key) {
    case '1':
    case 'b':
    case KEY_END:
    case '2':
    case 'j':
    case KEY_DOWN:
    case '3':
    case 'n':
    case KEY_NPAGE:
      if (d->io_offset[dim_y] != 72) {
        d->io_offset[dim_y] += 12;
      }
      break;
    case '4':
    case 'h':
    case KEY_LEFT:
    case '5':
    case ' ':
    case KEY_B2:
    case '6':
    case 'l':
    case KEY_RIGHT:
      break;
    case '7':
    case 'y':
    case KEY_HOME:
    case '8':
    case 'k':
    case KEY_UP:
    case '9':
    case 'u':
    case KEY_PPAGE:
      if (d->io_offset[dim_y]) {
        d->io_offset[dim_y] -= 12;
      }
      break;
    }
    switch (key) {
    case '1':
    case 'b':
    case KEY_END:
    case '4':
    case 'h':
    case KEY_LEFT:
    case '7':
    case 'y':
    case KEY_HOME:
      if (d->io_offset[dim_x]) {
        d->io_offset[dim_x] -= 40;
      }
      break;
    case '2':
    case 'j':
    case KEY_DOWN:
    case '5':
    case ' ':
    case KEY_B2:
    case '8':
    case 'k':
    case KEY_UP:
      break;
    case '3':
    case 'n':
    case KEY_NPAGE:
    case '6':
    case 'l':
    case KEY_RIGHT:
    case '9':
    case 'u':
    case KEY_PPAGE:
      if (d->io_offset[dim_x] != 80) {
        d->io_offset[dim_x] += 40;
      }
      break;
    }
    io_display(d);
  } while (1);
}

void io_handle_input(dungeon_t *d)
{
  uint32_t fail_code;
  int key;

  do {
    switch (key = getch()) {
    case '7':
    case 'y':
    case KEY_HOME:
      fail_code = move_pc(d, 7);
      break;
    case '8':
    case 'k':
    case KEY_UP:
      fail_code = move_pc(d, 8);
      break;
    case '9':
    case 'u':
    case KEY_PPAGE:
      fail_code = move_pc(d, 9);
      break;
    case '6':
    case 'l':
    case KEY_RIGHT:
      fail_code = move_pc(d, 6);
      break;
    case '3':
    case 'n':
    case KEY_NPAGE:
      fail_code = move_pc(d, 3);
      break;
    case '2':
    case 'j':
    case KEY_DOWN:
      fail_code = move_pc(d, 2);
      break;
    case '1':
    case 'b':
    case KEY_END:
      fail_code = move_pc(d, 1);
      break;
    case '4':
    case 'h':
    case KEY_LEFT:
      fail_code = move_pc(d, 4);
      break;
    case '5':
    case ' ':
    case KEY_B2:
      fail_code = 0;
      break;
    case '>':
      fail_code = move_pc(d, '>');
      break;
    case '<':
      fail_code = move_pc(d, '<');
      break;
    case 'L':
      io_look_mode(d);
      fail_code = 0;
      break;
    case 'S':
      d->save_and_exit = 1;
      d->pc.next_turn -= (1000 / d->pc.speed);
      fail_code = 0;
      break;
    case 'Q':
      /* Extra command, not in the spec.  Quit without saving.          */
      d->quit_no_save = 1;
      fail_code = 0;
      break;
    case 'H':
      /* Extra command, not in the spec.  H is for Huge: draw the whole *
       * dungeon, the pre-curses way.  Doesn't use a player turn.       */
      io_display_huge(d);
      fail_code = 1;
      break;
    default:
      /* Also not in the spec.  It's not always easy to figure out what *
       * key code corresponds with a given keystroke.  Print out any    *
       * unhandled key here.  Not only does it give a visual error      *
       * indicator, but it also gives an integer value that can be used *
       * for that key in this (or other) switch statements.  Printed in *
       * octal, with the leading zero, because ncurses.h lists codes in *
       * octal, thus allowing us to do reverse lookups.  If a key has a *
       * name defined in the header, you can use the name here, else    *
       * you can directly use the octal value.                          */
      mvprintw(0, 0, "Unbound key: %#o ", key);
      fail_code = 1;
    }
  } while (fail_code);
}
